Update documentation for KEP-3104 (#52877)
kubernetes/website: content/en/docs/reference/kubectl/kuberc.md

With this setting, running `kubectl delete pod/test-pod` will default to prompting for confirmation.
However, `kubectl delete pod/test-pod --interactive=false` will bypass the confirmation.

## Credential plugin policy

{{< feature-state for_k8s_version="v1.35" state="beta" >}}

Editors of a `kubeconfig` can specify an executable plugin that will be used to
acquire credentials to authenticate the client to the cluster. Within a `kuberc`
confguration, you can set the execution policy for such plugins by the use of
two top-level fields. Both fields are optional.

### credentialPluginPolicy

You can configure a policy for credentials plugins, using the optional
`credentialPluginPolicy` field. There are three valid values for this field:

1. `"AllowAll"`

When the policy is set to `"AllowAll"`, there will be no restrictions on which
plugins may run. This behavior is identical to that of Kubernetes versions prior
to 1.35.

2. `"DenyAll"`

When the policy is set to `"DenyAll"`, no exec plugins will be permitted to run.
If you don't use credential plugins, the Kubernetes project recommends that you
set the DenyAll policy. Using the Allowlist policy is a good choice if you do

> use credential plugins. If you are not sure, choose DenyAll.

Author: I ended up removing this section and kept the note below (in the only remaining place where the recommended default is discussed)

plugins. Each set of requirements will be attempted in turn; once the plugin
meets all requirement in at least one set, the plugin will be permitted to
execute. That is, the overall result of an application of the allowlist
to plugin `my-binary-plugin` is the _logical OR_ of the decisions rendered by
each item in the list.

As an example, consider the following allowlist configuration:

```yaml
apiVersion: kubectl.config.k8s.io/v1beta1
kind: Preference
credentialPluginPolicy: Allowlist
credentialPluginAllowlist:
  - name: foo
  - name: bar
  - name: baz
```

In the above example, the allowlist will allow plugins that have the name "foo",
"bar", _OR_ "baz".

Within each set of requirements, **all** specified (non-empty) requirements must be
met in order for the item to render an "allow" decision for the plugin. Put
another way, the decision rendered by an allowlist item is the _logical AND_ of
the decisions rendered by all nonempty fields on the item.
For Kubernetes {{< skew currentVersion >}}, the only
specifiable requirement is the `name` field (documented below).

To demonstrate the above, consider the following allowlist.
{{ <note> }}

> This allowlist is for illustration only, and will not actually work. As

Author: removed this example and the explanation it was illustrating.

mentioned above, the only specifiable requirement for Kubernetes
{{< skew currentVersion >}} is `name`. However, as fields are added it will be
important to understand how allowlist decisions are made.
{{ </note> }}

```yaml
apiVersion: kubectl.config.k8s.io/v1beta1
kind: Preference
credentialPluginPolicy: Allowlist
credentialPluginAllowlist:
  - name: foo

>     digest: "sha256:b9a3fad00d848ff31960c44ebb5f8b92032dc085020f857c98e32a5d5900ff9c"

Reviewer: +1 to not having fields that do not exist.

>     digest: "sha256:b9a3fad00d848ff31960c44ebb5f8b92032dc085020f857c98e32a5d5900ff9c"

Author: These were added because an example was requested specifically here. I thought it was a bit weird. I think the solution is to remove the explanation of future behavior (and this supporting example) as that is defined in the KEP and need not be in the docs until it's relevant.

4. is signed by the key "D4B35D8A0C68D4795242F442EA6287AC186B8334"

{{ <note> }}
For a set of requirements to be valid it MUST have at least one field that is
nonempty and explicitly specified. If all fields are empty or unspecified, it is
considered a configuration error and the plugin will not be allowed to execute.
Likewise if the `credentialPluginAllowlist` field is unspecified, or if it is
specified explicitly as the empty list. This is in order to prevent scenarios
where the user misspells the `credentialPluginAllowlist` key -- thinking they
have specified an allowlist when they actually haven't.

For example, the following is invalid:
```yaml
apiVersion: kubectl.config.k8s.io/v1beta1
kind: Preference
credentialPluginPolicy: Allowlist
credentialPluginAllowlist:
  - name: ""
```
{{ </note> }}

##### name

`name` names a credential plugin which may be executed. It can be specified as
either the basename of the desired plugin, or the full path. If specified as a
basename, the decision rendered by this field is "allow" if one of the following
two conditions is met:

1. The `name` field is exactly equal to the plugin's `command` field.

> 1. Full path resolution is performed on both the `name` and the `command`, and

Reviewer: Probably should mention that symlinks are not followed.

Reviewer 2: In what sense? If I make `/bin` a symlink, am I then unable to allowlist credential plugins with a path starting `/bin/`?

With this setting, running `kubectl delete pod/test-pod` will default to prompting for confirmation.
However, `kubectl delete pod/test-pod --interactive=false` will bypass the confirmation.

## Credential Plugin Policy

Editors of a `kubeconfig` can specify an executable plugin that will be used to
acquire credentials to authenticate the client to the cluster. Within a `kuberc`
confguration, you can set the execution policy for such plugins by the use of
two top-level fields. Both fields are optional.

### credentialPluginPolicy

The policy is configured using the `credentialPluginPolicy` field. There are
three valid values for this field:

1. `"AllowAll"`
1. `"DenyAll"`
1. `"Allowlist"`

**NOTE**: In order to maintain backward compatibility, an unspecified or empty
`credentialPluginPolicy` is identical to explicitly setting the policy to
`"AllowAll"`.

#### AllowAll

When the policy is set to `"AllowAll"`, there will be no restrictions on which
plugins may run. This behavior is identical to that of kubernetes versions prior
to 1.35.

> #### DenyAll

Reviewer 2: Try this: 
```suggestion
#### DenyAll {#credentialPluginPolicy-DenyAll}
```

It will avoid conflict if another field ever allows the same value.

When the policy is set to `"DenyAll"`, no exec plugins will be permitted to run.
This is the recommended default.

#### Allowlist

When the policy is set to `"Allowlist"`, the user can selectively allow
execution of credential plugins. When the policy is `"Allowlist"`, the user MUST
also provide the `credentialPluginAllowlist` field (also in the top-level). That
field is described below.

### credentialPluginAllowlist

The `credentialPluginAllowlist` field specifies a list of criteria-sets

> (henceforth called **requirements**) for permission to execute credential

Reviewer 2: ```suggestion
(sets of *requirements*) for permission to execute credential
```

Aim for a reading age of the median person who has completed mandatory education in English speaking countries.
That median person doesn't use "henceforth".

plugins. Each set of requirements will be attempted in turn; once the plugin
meets all requirement in at least one set, the plugin will be permitted to
execute. That is, the result overall result of an application of the allowlist
to plugin `my-binary-plugin` is the logical `OR` of the decisions rendered by
each item in the list.

*Within* each set of requirements, all specified/nonempty requirements must be
met in order for the item to render an "allow" decision for the plugin. Put
another way, the decision rendered by an allowlist item is the logical `AND` of

> the decisions rendered by all nonempty fields on the item. At present, the only

Reviewer 2: ```suggestion
the decisions rendered by all nonempty fields on the item.
For Kubernetes {{< skew currentVersion >}}, the only
```

specifiable requirement is the `name` field (documented below).

> Note that for a set of requirements to be valid it MUST have at least one field

Reviewer 2: Much better to use a [note shortcode](https://kubernetes.io/docs/contribute/style/style-guide/#note) here.

Also use boldface, not all-caps, for emphasis.

that is nonempty and explicitly specified. If all fields are empty or
unspecified, it is considered a configuration error and the plugin will not be
allowed to execute. Likewise if the `credentialPluginAllowlist` field is
unspecified, or if it is specified explicitly as the empty list. This is in
order to prevent scenarios where the user misspells the
`credentialPluginAllowlist` key -- thinking they have specified an allowlist
when they actually haven't.

##### name

`name` names a credential plugin which may be executed. It can be specified as
either the basename of the desired plugin, or the full path. If specified as a
basename, the decision rendered by this field is equivalent to the expression

> `name == plugin || exec.LookPath(name) == exec.LookPath(plugin)`. If specified

Reviewer 2: I don't think we c should expect readers to know the programming language described here. Ideally we'd rewrite it.

    options:
      - name: interactive
        default: "true"

> credentialPluginPolicy: DenyAll

Reviewer: We should add some verbiage that states that users of managed providers should consult the provider's documentation for what exec plugins may be needed in those environments.
